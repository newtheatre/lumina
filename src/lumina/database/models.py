import datetime
from enum import Enum
from typing import Optional

from fastapi.encoders import jsonable_encoder
from pydantic import BaseModel, EmailStr
from unit.lumina.util.types import BaseModelProtocol

from . import table


class DynamoExportMixin:
    def ddict(self: BaseModelProtocol, **kwargs):
        """dict() export for DynamoDB, removes any Python-only fields, essentially
        JSON but still in dict form"""
        return jsonable_encoder(self.dict(**kwargs))


class BaseDynamoModel(BaseModel):
    pk: str
    sk: str


class MemberConsent(BaseModel, DynamoExportMixin):
    consent_news: Optional[datetime.datetime]
    consent_network: Optional[datetime.datetime]
    consent_members: Optional[datetime.datetime]
    consent_students: Optional[datetime.datetime]


class MemberModel(BaseDynamoModel, DynamoExportMixin):
    sk = table.SK_PROFILE
    name: str
    email: EmailStr
    phone: Optional[str]
    year_of_graduation: Optional[int]
    created_at: Optional[datetime.datetime]
    email_verified_at: Optional[datetime.datetime]
    consent: Optional[MemberConsent]

    @property
    def id(self) -> str:
        return self.pk

    @property
    def email_verified(self) -> bool:
        return self.email_verified_at is not None

    def to_submitter(self) -> "SubmitterModel":
        return SubmitterModel(
            id=self.pk,
            verified=True,
            name=self.name,
            email=self.email,
            year_of_graduation=self.year_of_graduation,
        )


class SubmitterModel(BaseModel, DynamoExportMixin):
    id: str
    verified: bool
    name: str
    year_of_graduation: Optional[int]
    email: Optional[EmailStr]


class GitHubIssueState(Enum):
    OPEN = "open"
    CLOSED = "closed"


class GitHubIssueModel(BaseModel, DynamoExportMixin):
    number: int
    state: GitHubIssueState
    title: str
    created_at: datetime.datetime
    updated_at: datetime.datetime
    closed_at: Optional[datetime.datetime]
    comments: int


class SubmissionModel(BaseDynamoModel, DynamoExportMixin):
    url: str
    target_id: str
    target_type: str
    target_name: str
    created_at: datetime.datetime
    message: Optional[str]
    submitter: SubmitterModel
    github_issue: GitHubIssueModel

    @property
    def issue_id(self) -> int:
        # TODO: Use this when we have Python 3.9
        # return int(self.target_id.removeprefix(f"{table.SK_SUBMISSION}/"))
        return int(self.sk.split("/")[-1])
