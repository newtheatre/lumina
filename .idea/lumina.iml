<?xml version="1.0" encoding="UTF-8"?>
<module type="PYTHON_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/tests" isTestSource="true" />
      <excludeFolder url="file://$MODULE_DIR$/.aws-sam" />
      <excludeFolder url="file://$MODULE_DIR$/dependencies" />
      <excludeFolder url="file://$MODULE_DIR$/.pytest_cache" />
      <excludeFolder url="file://$MODULE_DIR$/.mypy_cache" />
      <excludeFolder url="file://$MODULE_DIR$/tests/.pytest_cache" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
  <component name="TemplatesService">
    <option name="TEMPLATE_CONFIGURATION" value="Jinja2" />
    <option name="TEMPLATE_FOLDERS">
      <list>
        <option value="$MODULE_DIR$/src/lumina/emails/templates" />
      </list>
    </option>
  </component>
</module>