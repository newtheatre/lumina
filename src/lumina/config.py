from pydantic_settings import BaseSettings


class Settings(BaseSettings):
    environment: str = "dev"
    stage_name: str = ""
    vcs_rev: str = "unknown"
    aws_region: str = "eu-west-2"

    github_owner: str = "newtheatre"
    github_repo: str = "lumina-test"

    sentry_dsn: str | None = None

    @property
    def root_path(self) -> str | None:
        if self.stage_name in {"", "dev", "Environment"}:
            return None
        if self.stage_name == "prod":
            return "https://lumina.nthp.wjdp.uk"
        return f"/{self.stage_name}"


settings = Settings()
