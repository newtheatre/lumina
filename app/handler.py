from mangum import Mangum

from lumina.app import app

handler = Mangum(app)
