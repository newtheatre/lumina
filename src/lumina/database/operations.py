from uuid import UUID

from boto3.dynamodb.conditions import Key
from lumina.database.models import GitHubIssueModel, MemberModel
from lumina.util import dates

from .models import SubmissionModel
from .table import (
    GSI_SK,
    GSI_SUBMISSION_TARGET,
    GSI_SUBMISSION_TARGET_PK,
    GSI_SUBMISSION_TARGET_SK,
    MEMBER_PARTITION_KEY,
    MEMBER_SORT_KEY,
    SK_PROFILE,
    SK_SUBMISSION_PREFIX,
    get_member_table,
    get_submission_sk,
)


class DbError(Exception):
    pass


class ResultNotFound(DbError):
    pass


def get_members() -> list[MemberModel]:
    options = {
        "IndexName": GSI_SK,
        "KeyConditionExpression": Key(MEMBER_SORT_KEY).eq(SK_PROFILE),
    }
    response = get_member_table().query(**options) # type: ignore
    data = response["Items"]
    while response.get("LastEvaluatedKey"):
        response = get_member_table().scan(
            ExclusiveStartKey=response["LastEvaluatedKey"],
            **options, # type: ignore
        )
        data.extend(response["Items"])
    return [MemberModel(**item) for item in data]


def get_member(id: str) -> MemberModel:
    response = get_member_table().get_item(
        Key={MEMBER_PARTITION_KEY: id, MEMBER_SORT_KEY: SK_PROFILE}
    )
    if result := response.get("Item"):
        return MemberModel(**result)
    raise ResultNotFound(f"Member with id {id} not found")


def put_member(model: MemberModel) -> MemberModel:
    get_member_table().put_item(Item=model.ddict())
    return model


def set_member_email_verified(id: str) -> None:
    get_member_table().update_item(
        Key={MEMBER_PARTITION_KEY: id, MEMBER_SORT_KEY: SK_PROFILE},
        UpdateExpression="set email_verified_at = :v",
        ExpressionAttributeValues={":v": dates.now().isoformat()},
    )


def delete_member(id: str) -> None:
    get_member_table().delete_item(
        Key={MEMBER_PARTITION_KEY: id, MEMBER_SORT_KEY: SK_PROFILE}
    )


def get_submission(id: int) -> SubmissionModel:
    response = get_member_table().query(
        IndexName=GSI_SK,
        KeyConditionExpression=Key(MEMBER_SORT_KEY).eq(get_submission_sk(id)),
    )
    if response["Count"] == 0:
        raise ResultNotFound(f"Submission with id {id} not found")
    if response["Count"] > 1:
        raise DbError(f"Multiple submissions with id {id} found")
    return SubmissionModel(**response["Items"][0])  # type: ignore


def get_submissions_for_member(id: str | UUID) -> list[SubmissionModel]:
    response = get_member_table().query(
        KeyConditionExpression=Key(MEMBER_PARTITION_KEY).eq(str(id))
        & Key(MEMBER_SORT_KEY).begins_with(SK_SUBMISSION_PREFIX),
    )
    return [SubmissionModel(**item) for item in response["Items"]]  # type: ignore


def get_submissions_for_target(
    target_type: str, target_id: str
) -> list[SubmissionModel]:
    response = get_member_table().query(
        IndexName=GSI_SUBMISSION_TARGET,
        KeyConditionExpression=Key(GSI_SUBMISSION_TARGET_PK).eq(target_type)
        & Key(GSI_SUBMISSION_TARGET_SK).eq(target_id),
    )
    return [SubmissionModel(**item) for item in response["Items"]]  # type: ignore


def put_submission(model: SubmissionModel) -> SubmissionModel:
    get_member_table().put_item(Item=model.ddict())
    return model


def update_submission_github_issue(id: int, issue: GitHubIssueModel) -> SubmissionModel:
    submission = get_submission(id)
    get_member_table().update_item(
        Key={
            MEMBER_PARTITION_KEY: submission.pk,
            MEMBER_SORT_KEY: get_submission_sk(id),
        },
        UpdateExpression="set github_issue = :v",
        ExpressionAttributeValues={":v": issue.ddict()},
    )
    return get_submission(id)


def move_anonymous_submissions_to_member(
    member_id: str, anonymous_id: UUID
) -> list[SubmissionModel]:
    """Delete all submissions for the anonymous member and create them as member
    submissions."""
    anonymous_submissions = get_submissions_for_member(anonymous_id)
    new_member_submissions = []
    for anonymous_submission in anonymous_submissions:
        # Create a new submission using the member ID
        new_member_submissions.append(
            # Direct copy and change only the PK
            put_submission(anonymous_submission.model_copy(update={"pk": member_id}))
        )
        # Delete the old submission
        delete_response = get_member_table().delete_item(
            Key={
                MEMBER_PARTITION_KEY: anonymous_submission.pk,
                MEMBER_SORT_KEY: anonymous_submission.sk,
            },
            ReturnValues="ALL_OLD",  # Needed for 'Attributes' in response
        )
        assert delete_response["Attributes"], "Original submission not deleted"
    return new_member_submissions
