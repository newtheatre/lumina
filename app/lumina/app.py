from fastapi import FastAPI

from lumina.config import settings
from lumina.router import router

app = FastAPI(title="Lumina", version=settings.vcs_rev, root_path=settings.root_path)
app.include_router(router)
