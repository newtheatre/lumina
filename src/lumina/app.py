from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from sentry_sdk.integrations.asgi import SentryAsgiMiddleware

import lumina.router
from lumina.config import settings

app = FastAPI(title="Lumina", version=settings.vcs_rev, root_path=settings.root_path)

app.add_middleware(SentryAsgiMiddleware)
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3000", "https://nthp-web.pages.dev"],
    allow_methods=["*"],
    allow_headers=["*"],
)

app.include_router(lumina.router.router)
lumina.router.use_route_names_as_operation_ids(app)
