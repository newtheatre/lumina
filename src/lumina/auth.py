from datetime import datetime, timedelta, timezone
from http import HTTPStatus
from typing import Optional

import jwt
from fastapi import Depends, HTTPException, Request
from fastapi.security import HTTPBearer
from pydantic import BaseModel

import lumina.database.operations
from lumina import ssm
from lumina.database.models import MemberModel

JWT_ALGORITHM = "RS256"
JWT_EXPIRATION_DELTA = timedelta(days=90)
AUTH_URL = "https://nthp-web.pages.dev/auth"


class AuthenticatedToken(BaseModel):
    id: str
    expires_at: datetime


def get_jwt_public_key() -> str:
    """Get public key for JWT verification."""
    return ssm.get_parameter("/lumina/jwt/public")


def get_jwt_private_key() -> str:
    """Get private key for JWT signing."""
    return ssm.get_parameter("/lumina/jwt/private")


def encode_jwt(sub: str) -> str:
    return jwt.encode(
        {
            "sub": sub,
            "exp": datetime.now(tz=timezone.utc) + JWT_EXPIRATION_DELTA,
        },
        get_jwt_private_key(),
        algorithm=JWT_ALGORITHM,
    )


def decode_jwt(token: str) -> AuthenticatedToken:
    payload = jwt.decode(token, get_jwt_public_key(), algorithms=[JWT_ALGORITHM])
    return AuthenticatedToken(id=payload["sub"], expires_at=payload["exp"])


class JWTBearer(HTTPBearer):
    def __init__(self, optional: bool = False):
        self.optional = optional
        super(JWTBearer, self).__init__(auto_error=False)

    async def __call__(self, request: Request) -> Optional[AuthenticatedToken]:
        credentials = await super(JWTBearer, self).__call__(request)
        if credentials:
            try:
                return decode_jwt(credentials.credentials)
            except jwt.InvalidTokenError:
                raise HTTPException(
                    status_code=HTTPStatus.UNAUTHORIZED,
                    detail="Invalid or expired token",
                )
        elif self.optional:
            return None
        else:
            raise HTTPException(
                status_code=HTTPStatus.UNAUTHORIZED, detail="Authorization required"
            )


def require_member(
    authenticated_member=Depends(JWTBearer()),
) -> MemberModel:
    """Get member if token is provided, otherwise raise exception."""
    try:
        return lumina.database.operations.get_member(authenticated_member.id)
    except lumina.database.operations.ResultNotFound:
        raise HTTPException(
            status_code=HTTPStatus.UNAUTHORIZED,
            detail="Member no longer exists",
        )


def optional_member(
    authenticated_member=Depends(JWTBearer(optional=True)),
) -> Optional[MemberModel]:
    """Get member if token is provided, otherwise return None."""
    if not authenticated_member:
        return None
    return require_member(authenticated_member)


def get_auth_url(sub: str) -> str:
    return f"{AUTH_URL}?token={encode_jwt(sub)}"
