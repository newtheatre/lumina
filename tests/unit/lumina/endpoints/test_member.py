from http import HTTPStatus
from unittest import mock

from fastapi.testclient import TestClient
from fixtures.models import MEMBER_MODEL_FRED_BLOGGS

from lumina.app import app
from lumina.database.operations import ResultNotFound

client = TestClient(app)

FAKE_TOKEN_URL = "https://nthp.test/auth?token=123"


class TestReadMember:
    def test_unauthorised(self):
        response = client.get("/member/fred_bloggs")
        assert response.status_code == HTTPStatus.UNAUTHORIZED

    def test_cannot_read_another_member(self, auth_fred_bloggs):
        response = client.get("/member/alice_bloggs")
        assert response.status_code == HTTPStatus.FORBIDDEN
        assert response.json() == {"detail": "You cannot read another member"}

    @mock.patch(
        "lumina.database.operations.get_member",
        return_value=MEMBER_MODEL_FRED_BLOGGS,
    )
    def test_success_self(self, mock_get_member, auth_fred_bloggs):
        response = client.get("/member/fred_bloggs")
        assert mock_get_member.called
        assert response.status_code == HTTPStatus.OK
        assert response.json() == {"id": "fred_bloggs", "email": "fred@bloggs.test"}


class TestCheckMember:
    @mock.patch(
        "lumina.database.operations.get_member",
        return_value=MEMBER_MODEL_FRED_BLOGGS,
    )
    def test_check_member_found(self, mock_get_member):
        response = client.get("/member/fred_bloggs/check")
        assert response.status_code == HTTPStatus.OK
        assert response.json() == {
            "id": "fred_bloggs",
            "maskedEmail": "fr***@bl***.test",
        }

    @mock.patch("lumina.database.operations.get_member", side_effect=ResultNotFound())
    def test_check_member_not_found(self, mock_get_member):
        response = client.get("/member/fred_bloggs/check")
        assert response.status_code == HTTPStatus.NOT_FOUND


class TestRegisterMember:
    @mock.patch("lumina.database.operations.create_member", return_value=None)
    @mock.patch("lumina.database.operations.get_member", side_effect=ResultNotFound())
    @mock.patch("lumina.auth.get_auth_url", return_value=FAKE_TOKEN_URL)
    @mock.patch("lumina.emails.send.send_email", return_value="abc123")
    def test_success(self, send_email, get_auth_url, get_member, create_member):
        response = client.post(
            "/member/fred_bloggs",
            json={
                "email": "test@example.com",
                "fullName": "Fred Bloggs",
            },
        )
        assert response.status_code == HTTPStatus.OK
        assert get_member.called
        assert send_email.called

    @mock.patch(
        "lumina.database.operations.get_member",
        return_value=MEMBER_MODEL_FRED_BLOGGS,
    )
    @mock.patch(
        "lumina.auth.get_auth_url", return_value="https://nthp.test/auth?token=123"
    )
    @mock.patch("lumina.emails.send.send_email", return_value="abc123")
    def test_conflict_member_already_exists(self, send_email, get_auth_url, get_member):
        response = client.post(
            "/member/test_id",
            json={
                "email": "test@example.com",
                "fullName": "Fred Bloggs",
            },
        )
        assert response.status_code == HTTPStatus.CONFLICT
        assert get_member.called
        assert not send_email.called

    def test_invalid_email(self):
        response = client.post(
            "/member/test_id",
            json={
                "email": "test@",
                "fullName": "Test Member",
            },
        )
        assert response.status_code == HTTPStatus.UNPROCESSABLE_ENTITY
        assert response.json() == {
            "detail": [
                {
                    "loc": ["body", "email"],
                    "msg": "value is not a valid email address",
                    "type": "value_error.email",
                }
            ]
        }


class TestDeleteMember:
    def test_no_auth(self):
        response = client.delete("/member/fred_bloggs")
        assert response.status_code == HTTPStatus.UNAUTHORIZED

    def test_cannot_delete_others(self, auth_fred_bloggs):
        response = client.delete("/member/alice_bloggs")
        assert response.status_code == HTTPStatus.FORBIDDEN
        assert response.json() == {"detail": "You cannot delete another member"}

    @mock.patch("lumina.database.operations.delete_member", return_value=None)
    def test_success(self, mock_delete_member, auth_fred_bloggs):
        response = client.delete("/member/fred_bloggs")
        assert response.status_code == HTTPStatus.OK
        assert mock_delete_member.called


class TestSendTokenLinkForMember:
    @mock.patch("lumina.database.operations.get_member", side_effect=ResultNotFound())
    def test_no_member(self, get_member):
        response = client.post("/member/fred_bloggs/login")
        assert get_member.called
        assert response.status_code == HTTPStatus.NOT_FOUND

    @mock.patch("lumina.auth.get_auth_url", return_value=FAKE_TOKEN_URL)
    @mock.patch("lumina.emails.send.send_email", return_value="abc123")
    @mock.patch(
        "lumina.database.operations.get_member", return_value=MEMBER_MODEL_FRED_BLOGGS
    )
    def test_success(self, get_member, send_email, get_auth_url):
        response = client.post("/member/fred_bloggs/login")
        assert get_member.called
        assert send_email.called
        assert get_auth_url.called
        assert response.status_code == HTTPStatus.OK
