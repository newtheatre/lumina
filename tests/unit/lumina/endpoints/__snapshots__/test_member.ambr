# serializer version: 1
# name: TestReadMember.test_do_not_verify_if_already_verified
  dict({
    'anonymousIds': None,
    'consent': dict({
      'consentMembers': False,
      'consentNetwork': False,
      'consentNews': False,
      'consentStudents': False,
    }),
    'createdAt': None,
    'email': 'fred@bloggs.com',
    'emailVerified': True,
    'emailVerifiedAt': '2021-01-01T00:00:00',
    'id': 'fred_bloggs',
    'phone': None,
    'yearOfGraduation': None,
  })
# ---
# name: TestReadMember.test_move_anonymous_submissions
  dict({
    'anonymousIds': list([
      '5f72853b-1cab-4e1c-98e2-bbff6b0cec5b',
    ]),
    'consent': dict({
      'consentMembers': False,
      'consentNetwork': False,
      'consentNews': False,
      'consentStudents': False,
    }),
    'createdAt': None,
    'email': 'fred@bloggs.com',
    'emailVerified': True,
    'emailVerifiedAt': '2021-01-01T00:00:00',
    'id': 'fred_bloggs',
    'phone': None,
    'yearOfGraduation': None,
  })
# ---
# name: TestReadMember.test_success_self_first_call
  dict({
    'anonymousIds': None,
    'consent': dict({
      'consentMembers': False,
      'consentNetwork': False,
      'consentNews': False,
      'consentStudents': False,
    }),
    'createdAt': None,
    'email': 'fred@bloggs.com',
    'emailVerified': True,
    'emailVerifiedAt': '2021-01-01T00:00:00',
    'id': 'fred_bloggs',
    'phone': None,
    'yearOfGraduation': None,
  })
# ---
# name: TestUpdateMember.test_success
  _Call(
    tuple(
      MemberModel(pk='fred_bloggs', sk='profile', name='Fred Bloggs', email='fred@bloggs.com', phone='01234567890', year_of_graduation=None, created_at=None, email_verified_at=None, consent=MemberConsentModel(consent_news=FakeDatetime(2021, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), consent_network=FakeDatetime(2021, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), consent_members=FakeDatetime(2021, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), consent_students=FakeDatetime(2021, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)), anonymous_ids=None),
    ),
    dict({
    }),
  )
# ---
